test/parallel/test-stream-pipeline.js

```diff
@@ -0,0 +1,244 @@
+    read() {}
+  });
+
+  const transform = new Transform({
+    transform(data, enc, cb) {
+      cb(new Error('kaboom'));
+    }
+  });
+
+  const write = new Writable({
+    write(data, enc, cb) {
+      cb();
+    }
+  });
+
+  read.on('close', common.mustCall());
+  transform.on('close', common.mustCall());
+  write.on('close', common.mustCall());
+
+  pipeline(read, transform, write, common.mustCall((err) => {
+    assert.deepStrictEqual(err, new Error('kaboom'));
+  }));
+
+  read.push('hello');
+}
+
+{
+  const server = http.createServer((req, res) => {
+    const rs = new Readable({
+      read() {
+        rs.push('hello');
+        rs.push(null);
+      }
+    });
+
+    pipeline(rs, res);
+  });
+
+  server.listen(0, () => {
+    const req = http.request({
+      port: server.address().port
+    });
+
+    req.end();
+    req.on('response', (res) => {
+      const buf = [];
+      res.on('data', (data) => buf.push(data));
+      res.on('end', common.mustCall(() => {
+        assert.deepStrictEqual(
+          Buffer.concat(buf),
+          Buffer.from('hello')
+        );
+        server.close();
+      }));
+    });
+  });
+}
+
+{
+  const server = http.createServer((req, res) => {
+    const rs = new Readable({
+      read() {
+        rs.push('hello');
+      },
+      destroy: common.mustCall((err, cb) => {
+        // prevents fd leaks by destroying http pipelines
+        cb();
+      })
+    });
+
+    pipeline(rs, res);
+  });
+
+  server.listen(0, () => {
+    const req = http.request({
+      port: server.address().port
+    });
+
+    req.end();
+    req.on('response', (res) => {
+      setImmediate(() => {
+        res.destroy();
+        server.close();
+      });
+    });
+  });
+}
+
+{
+  const server = http.createServer((req, res) => {
+    const rs = new Readable({
+      read() {
+        rs.push('hello');
+      },
+      destroy: common.mustCall((err, cb) => {
+        cb();
+      })
+    });
+
+    pipeline(rs, res);
+  });
+
+  let cnt = 10;
+
+  const badSink = new Writable({
+    write(data, enc, cb) {
+      cnt--;
+      if (cnt === 0) cb(new Error('kaboom'));
+      else cb();
+    }
+  });
+
+  server.listen(0, () => {
+    const req = http.request({
+      port: server.address().port
+    });
+
+    req.end();
+    req.on('response', (res) => {
+      pipeline(res, badSink, common.mustCall((err) => {
+        assert.deepStrictEqual(err, new Error('kaboom'));
+        server.close();
+      }));
+    });
+  });
+}
+
+{
+  const pipelinePromise = promisify(pipeline);
+
+  async function run() {
+    const read = new Readable({
+      read() {}
+    });
+
+    const write = new Writable({
+      write(data, enc, cb) {
+        cb();
+      }
+    });
+
+    read.push('data');
+    read.push(null);
+
+    let finished = false;
+
+    write.on('finish', () => {
+      finished = true;
+    });
+
+    await pipelinePromise(read, write);
+
+    // nexttick to avoid the promise catching it
```

Comment:
this nextTick is not clear to me.

Reply:
was to work around the need for `common.crashOnUnhandledRejection()` which i didn't know about :+1: